src/cargo/core/compiler/future_incompat.rs

```diff
@@ -24,13 +52,239 @@ pub struct Diagnostic {
+            }
+        };
+        let report = OnDiskReport {
+            id: current_reports.next_id,
+            report: render_report(ws, per_package_reports),
+        };
+        current_reports.next_id += 1;
+        current_reports.reports.push(report);
+        if current_reports.reports.len() > MAX_REPORTS {
+            current_reports.reports.remove(0);
+        }
+        let on_disk = serde_json::to_vec(&current_reports).unwrap();
+        if let Err(e) = ws
+            .target_dir()
+            .open_rw(
+                FUTURE_INCOMPAT_FILE,
+                ws.config(),
+                "Future incompatibility report",
+            )
+            .and_then(|file| {
+                let mut file = file.file();
+                file.set_len(0)?;
+                file.write_all(&on_disk)?;
+                Ok(())
+            })
+        {
+            crate::display_warning_with_error(
+                "failed to write on-disk future incompatible report",
+                &e,
+                &mut ws.config().shell(),
+            );
+        }
+        current_reports
+    }
+
+    /// Loads the on-disk reports.
+    pub fn load(ws: &Workspace<'_>) -> CargoResult<OnDiskReports> {
+        let report_file = match ws.target_dir().open_ro(
+            FUTURE_INCOMPAT_FILE,
+            ws.config(),
+            "Future incompatible report",
+        ) {
+            Ok(r) => r,
+            Err(e) => {
+                if let Some(io_err) = e.downcast_ref::<std::io::Error>() {
+                    if io_err.kind() == std::io::ErrorKind::NotFound {
+                        bail!("no reports are currently available");
+                    }
+                }
+                return Err(e);
+            }
+        };
+
+        let mut file_contents = String::new();
+        report_file
+            .file()
+            .read_to_string(&mut file_contents)
+            .with_context(|| "failed to read report")?;
+        let on_disk_reports: OnDiskReports =
+            serde_json::from_str(&file_contents).with_context(|| "failed to load report")?;
+        if on_disk_reports.version != 0 {
+            bail!("unable to read reports; reports were saved from a future version of Cargo");
+        }
+        Ok(on_disk_reports)
+    }
+
+    /// Returns the most recent report ID.
+    pub fn last_id(&self) -> u32 {
+        self.reports.last().map(|r| r.id).unwrap()
+    }
+
+    pub fn get_report(&self, id: u32, config: &Config) -> CargoResult<String> {
+        let report = self.reports.iter().find(|r| r.id == id).ok_or_else(|| {
+            let available = iter_join(self.reports.iter().map(|r| r.id.to_string()), ", ");
+            format_err!(
+                "could not find report with ID {}\n\
+                 Available IDs are: {}",
+                id,
+                available
+            )
+        })?;
+        let report = if config.shell().err_supports_color() {
+            report.report.clone()
```

Comment:
I forget the exact specifics here, but is this the right thing to do on Windows? I thought we had to use a special crate which, while printing, interpreted ansi color codes into the windows console commands?

Reply:
Ah, sorry, I completely forgot.  Yea, it should be using `fwdansi` to print. I have updated it.